dotCMS/src/main/java/com/dotcms/auth/providers/saml/v1/SAMLHelper.java

```diff
@@ -0,0 +1,469 @@
+            user                   = Company.AUTH_TYPE_ID.equals(authType)?
+                    this.userAPI.loadUserById(this.samlAuthenticationService.getValue(attributes.getNameID()),      systemUser, false):
+                    this.userAPI.loadByUserByEmail(this.samlAuthenticationService.getValue(attributes.getNameID()), systemUser, false);
+        } catch (NoSuchUserException e) {
+
+            Logger.error(this, "No user matches ID '" +
+                    this.samlAuthenticationService.getValue(attributes.getNameID()) + "'. Creating one...", e);
+            user = null;
+        } catch (Exception e) {
+
+            Logger.error(this, "An error occurred when loading user with ID '" +
+                    (null != attributes && null != attributes.getNameID()?
+                    this.samlAuthenticationService.getValue(attributes.getNameID()): "null") + "'", e);
+            user = null;
+        }
+
+        // check if the client wants synchronization
+        final boolean createUserWhenDoesNotExists = DotSamlProxyFactory.getInstance()
+                .samlConfigurationService().getConfigAsBoolean(identityProviderConfiguration, SamlName.DOT_SAML_ALLOW_USER_SYNCHRONIZATION);
+        if (createUserWhenDoesNotExists) {
+
+            user = null == user?
+                    this.createNewUser(systemUser,    attributes, identityProviderConfiguration):  // if user does not exists, create a new one.
+                    this.updateUser(user, systemUser, attributes, identityProviderConfiguration); // update it, since exists
+
+            if (user.isActive()) {
+
+                this.addRoles(user, attributes, identityProviderConfiguration);
+            } else {
+
+                Logger.info(this, ()-> "User with ID '" + this.samlAuthenticationService.getValue(attributes.getNameID()) + "' is not active. No roles " +
+                        "were added.");
+            }
+        }
+
+        return user;
+    }
+
+    protected User updateUser(final User user, final User systemUser,
+                              final Attributes attributesBean, final IdentityProviderConfiguration identityProviderConfiguration) {
+        try {
+
+            if (DotSamlProxyFactory.getInstance().samlConfigurationService()
+                    .getConfigAsBoolean(identityProviderConfiguration, SamlName.DOTCMS_SAML_LOGIN_UPDATE_EMAIL)){
+
+                user.setEmailAddress(attributesBean.getEmail());
+            }
+
+            user.setFirstName(attributesBean.getFirstName());
+            user.setLastName(attributesBean.getLastName());
+
+            this.userAPI.save(user, systemUser, false);
+            Logger.debug(this, ()-> "User with email '" + attributesBean.getEmail() + "' has been updated");
+        } catch (Exception e) {
+
+            Logger.error(this, "Error updating user with email '" + attributesBean.getEmail() + "': " + e.getMessage()
+                    , e);
+            throw new SamlException(e.getMessage());
```

Comment:
Codacy found an issue: [New exception is thrown in catch block, original stack trace may be lost](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)